upcoming/ssh-server-in-go.md

```diff
@@ -0,0 +1,240 @@
+and just work. It will make this post too long if involves all the details.
+
+### Prepare to start a SSH server
+
+The server must have a private key in order to start a SSH server. This is for the
+purpose of preventing [Man-in-the-middle attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack).
+
+This key does not need to be server-wide, just keep it somewhere but not in temporary
+directory because users will add this key to their `known_hosts` file.
+
+```go
+// Listen starts a SSH server listens on given port.
+func Listen(port int) {
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			pkey, err := models.SearchPublicKeyByContent(strings.TrimSpace(string(ssh.MarshalAuthorizedKey(key))))
+			if err != nil {
+                // handle error
+				return nil, err
+			}
+			return &ssh.Permissions{Extensions: map[string]string{"key-id": com.ToStr(pkey.ID)}}, nil
+		},
+	}
+
+	keyPath := filepath.Join(setting.AppDataPath, "ssh/gogs.rsa")
+	if !com.IsExist(keyPath) {
+		os.MkdirAll(filepath.Dir(keyPath), os.ModePerm)
+		_, stderr, err := com.ExecCmd("ssh-keygen", "-f", keyPath, "-t", "rsa", "-N", "")
+		if err != nil {
+			panic(fmt.Sprintf("Fail to generate private key: %v - %s", err, stderr))
+		}
+		log.Trace("New private key is generateed: %s", keyPath)
+	}
+
+	privateBytes, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		panic("Fail to load private key")
+	}
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		panic("Fail to parse private key")
+	}
+	config.AddHostKey(private)
+
+	go listen(config, port)
+}
+```
+
+This piece of code does three things:
+
+1. Setup a callback for validating public key from database.
+
+    Function `ssh.MarshalAuthorizedKey` will return a string format of user's public
+    key with a line break, so we want to remove that by calling `strings.TrimSpace`,
+    and then search in the database.
+
+    After search, if we return any kind of error, it will produce `Permission denied`
+    prompt on user side. If no error is returned, you can carry an instance of type
+    `*ssh.Permissions` to the corresponding request handler.
+
+    In this case, we need to set which key ID is this request corresponding to in `Extensions`.
+
+2. Create a private key when there is no one exists.
+
+    This is done by calling a command `ssh-keygen -f keypath -t rsa -N ""`.
+
+3. Load private key and start listening on given port.
+
+### Start listening and accepting new requests
+
+Like normal HTTP server, an SSH server needs to listen on a specific port as well.
+
+The pattern is very similar:
+
+```go
+func listen(config *ssh.ServerConfig, port int) {
+	listener, err := net.Listen("tcp", "0.0.0.0:"+com.ToStr(port))
+	if err != nil {
+		panic(err)
+	}
+	for {
+		// Once a ServerConfig has been configured, connections can be accepted.
+		conn, err := listener.Accept()
+		if err != nil {
+            // handle error
```

Comment:
The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.